test/integration/ha_test.go

```diff
@@ -0,0 +1,627 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/minikube/cmd/minikube/cmd"
+	"k8s.io/minikube/pkg/minikube/config"
+	"k8s.io/minikube/pkg/util/retry"
+)
+
+// TestHA tests all ha cluster functionality
+func TestHA(t *testing.T) {
+	if NoneDriver() {
+		t.Skip("none driver does not support multinode/ha")
+	}
+
+	if DockerDriver() {
+		rr, err := Run(t, exec.Command("docker", "version", "-f", "{{.Server.Version}}"))
+		if err != nil {
+			t.Fatalf("docker is broken: %v", err)
+		}
+		if strings.Contains(rr.Stdout.String(), "azure") {
+			t.Skip("kic containers are not supported on docker's azure")
+		}
+	}
+
+	type validatorFunc func(context.Context, *testing.T, string)
+	profile := UniqueProfileName("ha")
+	ctx, cancel := context.WithTimeout(context.Background(), Minutes(30))
+	defer CleanupWithLogs(t, profile, cancel)
+
+	t.Run("serial", func(t *testing.T) {
+		tests := []struct {
+			name      string
+			validator validatorFunc
+		}{
+			{"StartCluster", validateHAStartCluster},
+			{"DeployApp", validateHADeployApp},
+			{"PingHostFromPods", validateHAPingHostFromPods},
+			{"AddWorkerNode", validateHAAddWorkerNode},
+			{"NodeLabels", validateHANodeLabels},
+			{"HAppyAfterClusterStart", validateHAStatusHAppy},
+			{"CopyFile", validateHACopyFile},
+			{"StopSecondaryNode", validateHAStopSecondaryNode},
+			{"DegradedAfterControlPlaneNodeStop", validateHAStatusDegraded},
+			{"RestartSecondaryNode", validateHARestartSecondaryNode},
+			{"HAppyAfterSecondaryNodeRestart", validateHAStatusHAppy},
+			{"RestartClusterKeepsNodes", validateHARestartClusterKeepsNodes},
+			{"DeleteSecondaryNode", validateHADeleteSecondaryNode},
+			{"DegradedAfterSecondaryNodeDelete", validateHAStatusDegraded},
+			{"StopCluster", validateHAStopCluster},
+			{"RestartCluster", validateHARestartCluster},
+			{"DegradedAfterClusterRestart", validateHAStatusDegraded},
+			{"AddSecondaryNode", validateHAAddSecondaryNode},
+			{"HAppyAfterSecondaryNodeAdd", validateHAStatusHAppy},
+		}
+		for _, tc := range tests {
+			tc := tc
+			if ctx.Err() == context.DeadlineExceeded {
+				t.Fatalf("Unable to run more tests (deadline exceeded)")
+			}
+			t.Run(tc.name, func(t *testing.T) {
+				defer PostMortemLogs(t, profile)
+				tc.validator(ctx, t, profile)
+			})
+		}
+	})
+}
+
+// validateHAStartCluster ensures ha cluster can start.
```

Comment:
lets add Mutli Control plane in parantehes for anywhere HA is specified.

Reply:
@medyagh i've added it